### ISSUE TITLE
Repo Consolidation: Round 3

### ISSUE THREAD

**Issue by marten-seemann** [libp2p/go-libp2p#1556]
The transport consolidation (#1187) has dramatically reduced repo sprawl. We now only have a few go-libp2p-* repos left in the dependency tree of go-lipbp2p. This is a proposal to also consolidate them into go-lipbp2:

- [x] go-libp2p-resource-manager: `/p2p/host/resource-manager`
   - [x] Update links in dos-mitigation.md: https://github.com/libp2p/docs/pull/160/
- [x] go-peerstore: `/p2p/host/peerstore`
- [x] go-libp2p-core: `/core`
- [x] go-eventbus: `/p2p/host/eventbus`

Not sure what to do about go-libp2p-asn-util. This repo should probably have been named go-asn-util, it's not libp2p related in any way. I also don't feel like moving several MB of ASN here.

Anything I missed? Thoughts, @MarcoPolo @vyzo @Stebalien @raulk?

**Comment by vyzo** [comment]
asn no, the rest is reasonable proposition.

**Comment by vyzo** [comment]
I think we should keep core separate, it's the interface after all.

**Comment by marten-seemann** [comment]
> I think we should keep core separate, it's the interface after all.

I think I disagree. core being separate is causing us a lot of trouble:
* When we want to change one of the interfaces, we first have to cut a core release before we can use them in go-libp2p. It would be nicer to make that change atomically.
* core and go-libp2p being out of sync (core already having released an update, but go-libp2p not yet) causes errors for downstream users who run `go get -u`, and then we need to tell them which versions of core go with which versions of go-libp2p

We probably should keep the GitHub Action that posts the `gocompat` output every time we make a change to `/core`, so we don't accidentally break things.

**Comment by vyzo** [comment]
Yes, but it allows libraries to just depend on core and not pull the whole
go-libp2p except for trsts.

**Comment by marten-seemann** [comment]
Should we be worried about this? I'd assume that if you pull in libp2p types, you're also using go-libp2p. You _could_ imagine a project suffering from repo sprawl where a library only needs the type, and go-libp2p is pulled in by their main repo, but even then the only effect is that `go get` will download slightly more data.

**Comment by vyzo** [comment]
It is semantically different; i think a separate interface is good practice for something as complex as libp2p.
After all, apart from instantiation and tests, user code only cares about the interface.

**Comment by vyzo** [comment]
think of it as the header file.

**Comment by marten-seemann** [comment]
Any suggestion how to fix the two problems I listed above then? For me, this is mostly about development velocity, and I want to get away from having to release one module (in a release that we _don't_ want users to use yet!) in order to make progress in another.

**Comment by vyzo** [comment]
Any changes in interface should be carefully considered and too much velocity there is indicative of instability.
We dont have to release before creating the implementation pr, and only release when the implementation in go-libp2p is ready to merge.
The window of a broken go get -u is insignificantly small then.

**Comment by marten-seemann** [comment]
Not convinced of the cleanliness argument, but let's look at some real-world examples. I looked at libraries building on top of libp2p, and determined their  libp2p dependencies after deleting all test files (`rm **/*_test.go && go mod edit -go 1.17 && go mod tidy`).

### go-libp2p-pubsub

```
github.com/libp2p/go-libp2p-core v0.15.1
github.com/libp2p/go-libp2p-discovery v0.6.0
```
go-libp2p-discovery was moved into go-libp2p, so this would still need to import go-libp2p.

### go-libp2p-kad-dht

```
github.com/libp2p/go-eventbus v0.2.1
github.com/libp2p/go-libp2p-core v0.15.1
github.com/libp2p/go-libp2p-kbucket v0.4.7
github.com/libp2p/go-libp2p-peerstore v0.6.0
github.com/libp2p/go-libp2p-record v0.1.3
github.com/libp2p/go-libp2p-routing-helpers v0.2.3
github.com/libp2p/go-libp2p-swarm v0.10.2
github.com/libp2p/go-libp2p-xor v0.1.0
github.com/libp2p/go-msgio v0.2.0
github.com/libp2p/go-netroute v0.2.0
```

go-libp2p-swarm (moved into go-libp2p in round 2) is imported for an error assertion, and go-libp2p-peerstore (will be moved in round 3) for the slightly weird function [`PeerInfos`](https://pkg.go.dev/github.com/libp2p/go-libp2p-peerstore#PeerInfos).

### go-bitswap

```
github.com/libp2p/go-buffer-pool v0.0.2
github.com/libp2p/go-libp2p v0.14.3
github.com/libp2p/go-libp2p-core v0.8.5
github.com/libp2p/go-libp2p-loggables v0.1.0
```
go-libp2p is import to access the ping package.

### go-graphsync

```
github.com/libp2p/go-buffer-pool v0.0.2
github.com/libp2p/go-libp2p-core v0.8.5
github.com/libp2p/go-msgio v0.0.6
```

This actually wouldn't need to import go-libp2p.

**Comment by MarcoPolo** [comment]
I think we can still be very deliberate with changes to the core interfaces even if it's in a single repo. I'm not aware of many libraries that have their interfaces defined in a separate repo, but please correct me if I'm wrong.

From a user's POV a single repo would be better since go-libp2p + interfaces are well defined for any given commit.

**Comment by vyzo** [comment]
My main rationale for the interface being a srparate rrpo is that libp3p is a very complex system, and that applications and libraries only ever care about the core interface (except for instantiation and perhaps tests).

I don't want our users to get in the habbit of using internal libp2p components; the libp2p contract is that the interface is stable, and in internals anything goes.

The development velocity effect is noe gone, given rhat we are essentially a monorepo.
When we want to make a change in core, we open pr, develop uaing that pr in the monorep, and when its ready we merge both.

No release is necessary in eithe repo, it is totally fine for master to point to core master.
When we are ready for rrlease, we bump both and off we go. The probablity of a user soing a `go get -u` and brealing while wr are tagging releases is vanishingly small.

**Comment by BigLep** [comment]
A couple of thoughts:
1. ResourceManager readme docs in https://github.com/libp2p/docs/pull/160/ will need to be updated with new links when this is done.
2. If it isn't there already, can we please capture the repo/package rationale in the README (likely taking content from https://github.com/libp2p/go-libp2p/issues/1556#issuecomment-1143924221 )?

**Comment by p-shahi** [comment]
Closing as "Update links in dos-mitigation.md" was addressed in libp2p/docs#165
